pkg/queue/concurrency_state_test.go

```diff
@@ -0,0 +1,197 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	"go.uber.org/atomic"
+	pkglogging "knative.dev/pkg/logging"
+
+	network "knative.dev/networking/pkg"
+)
+
+func TestConcurrencyStateHandler(t *testing.T) {
+	tests := []struct {
+		name            string
+		pauses, resumes int64
+		events          map[time.Duration]time.Duration // start time => req length
+	}{{
+		name:    "single request",
+		pauses:  1,
+		resumes: 1,
+		events: map[time.Duration]time.Duration{
+			1 * time.Second: 2 * time.Second,
+		},
+	}, {
+		name:    "overlapping requests",
+		pauses:  1,
+		resumes: 1,
+		events: map[time.Duration]time.Duration{
+			25 * time.Millisecond: 100 * time.Millisecond,
+			75 * time.Millisecond: 200 * time.Millisecond,
+		},
+	}, {
+		name:    "subsumbed request",
+		pauses:  1,
+		resumes: 1,
+		events: map[time.Duration]time.Duration{
+			25 * time.Millisecond: 300 * time.Millisecond,
+			75 * time.Millisecond: 200 * time.Millisecond,
+		},
+	}, {
+		name:    "start stop start",
+		pauses:  2,
+		resumes: 2,
+		events: map[time.Duration]time.Duration{
+			25 * time.Millisecond:  300 * time.Millisecond,
+			75 * time.Millisecond:  200 * time.Millisecond,
+			850 * time.Millisecond: 300 * time.Millisecond,
+			900 * time.Millisecond: 400 * time.Millisecond,
+		},
+	}}
+
+	logger, _ := pkglogging.NewLogger("", "error")
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			paused := atomic.NewInt64(0)
+			pause := func() {
+				paused.Inc()
+			}
+
+			resumed := atomic.NewInt64(0)
+			resume := func() {
+				resumed.Inc()
+			}
+
+			delegated := atomic.NewInt64(0)
+			delegate := func(w http.ResponseWriter, r *http.Request) {
+				wait, err := strconv.Atoi(r.Header.Get("wait"))
+				if err != nil {
+					panic(err)
+				}
+
+				time.Sleep(time.Duration(wait))
+				delegated.Inc()
+			}
+
+			h := ConcurrencyStateHandler(logger, http.HandlerFunc(delegate), pause, resume)
+
+			var wg sync.WaitGroup
+			wg.Add(len(tt.events))
+			for delay, length := range tt.events {
+				length := length
+				time.AfterFunc(delay, func() {
+					w := httptest.NewRecorder()
+					r := httptest.NewRequest("GET", "http://target", nil)
+					r.Header.Set("wait", strconv.FormatInt(int64(length), 10))
+					h.ServeHTTP(w, r)
+					wg.Done()
+				})
+			}
+
+			wg.Wait()
+			// Allow last update to finish (otherwise values are off, though this doesn't show
+			// as a race condition when running `go test -race `
+			// TODO Less hacky fix for this
+			time.Sleep(100 * time.Microsecond)
```

Comment:
Perhaps just poll until timeout or condition satisfied as we do in most other places?

Reply:
This is probably a silly question, but what would the condition function be in this case?